keps/sig-node/2823-pod-addmission-plugin/README.md

```diff
@@ -0,0 +1,445 @@
+will enable cloud providers who support nodeless/serverless worker nodes to
+restrict Pods based on fields other than those already being validated by
+kubelet.
+
+Also, Kubernetes scheduler is extensible but isn’t omniscient about everything
+on the node hence there can be cases where a scheduling decision cannot be
+fulfilled by the node based on the current node conditions such as node-level
+resource allocation. The in-tree PodAdmitHandlers mentioned below are good
+examples for this.
+
+* Eviction manager rejects pods if the node is under high pressure (e.g., PIDs)
+* Topology manager rejects pods if the NUMA affinity cannot be satisfied
+
+There can be additional resource allocation conditions and policies handled
+externally to the kubelet, e.g.,
+
+* Alternative resource allocator implemented by the CRI which needs to reject
+  pods in conditions similar to the built-in resource managers.
+
+These can also benefit from a pluggable admission handler to reject pods before
+they are admitted by the kubelet. This is hard to be replicated by exposing more
+information to the scheduler which may be too costly and subject to race
+conditions.
+
+### Goals
+
+- Allow deployers of fully managed worker nodes to have control over Pods
+  running on those nodes.
+
+- Enable node-local Pod admission checks without requiring changes to kubelet.
+
+### Non-Goals
+
+- Move existing validations to “out of tree” plugins.
+
+- Change existing API-layer validation solutions such as Pod Security Policies
+  and validating admission webhooks.
+
+## Proposal
+
+The approach taken is similar to the container networking interface (CNI) plugin
+architecture. With CNI, kubelet invokes one or more CNI plugin binaries on the
+host to set up a Pod’s networking. kubelet discovers available CNI plugins
+by [examining](https://github.com/kubernetes/kubernetes/blob/dd5272b76f07bea60628af0bb793f3cca385bf5e/pkg/kubelet/dockershim/docker_service.go#L242)
+a well-known directory (`/etc/cni/net.d`) for configuration files
+and [loading](https://github.com/kubernetes/kubernetes/blob/dd5272b76f07bea60628af0bb793f3cca385bf5e/pkg/kubelet/dockershim/docker_service.go#L248)
+plugin [descriptors](https://github.com/kubernetes/kubernetes/blob/f4db8212be53c69a27d893d6a4111422fbce8008/pkg/kubelet/dockershim/network/plugins.go#L52)
+upon startup.
+
+To support pluggable validation for pod admission on the worker node, we propose
+to have kubelet similarly discover node-local Pod admission plugins listed in a
+new PodAdmissionPluginDir flag.
+
+Other option is to enable this functionality through feature flag
+“enablePodAdmissionPlugin” and have the directory path defined inside the
+kubelet itself.
+
+### User Stories (Optional)
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+These out-of-tree plugins should be provisioned and managed along with kubelet.
```

Comment:
Let's capture as part of this KEP, from a sys-admin's perspective where would these out-of-tree plugins be would consumed from? Do we want to have a common repository in https://github.com/kubernetes-sigs to house all the out-of-tree admission plugins. Examples of repos for scheduler plugins and CNI plugins are below:

-  scheduler plugins: https://github.com/kubernetes-sigs/scheduler-plugins
-  CNI plugins: https://github.com/containernetworking/plugins